### PR TITLE
docs(legacy): added legacy documentation that was removed earlier

### LINKED ISSUE #2718: Restore consensus-reactor.md
## Bug Report
The `spec/reactors` folder was deleted back in 2021 in [this](https://github.com/tendermint/tendermint/commit/72d15a4b0773adb8d5160a34d9a8e2d981394484#diff-5531f9389977e6bcb98535af7a345e3d7d81d74fd53e726357a1a7b441bc12fb) commit. I'm not sure yet why; my hypothesis is that the information from those folders was disseminated into other documentation.

However, the content of `spec/reactors/consensus/consensus-reactor.md` doesn't seem to be available anywhere. I've checked the other documentation folders, the CometBFT documentation website, and the Tendermint documentation website.

It seems that reactor documentation and tutorials have been sparse lately.

Should we restore the whole `spec/reactors` folder or only the `consunsus-reactor.md` file?

Maybe there is a good reason why the folder was deleted and we don't need them anymore.

### PR DESCRIPTION
**Pull request by greg-szabo**
Refs #2718 

Revert "spec: remove reactor section (#242)" Restore spec/reactors folder by reverting changes in commit 72d15a4

This partially reverts commit 72d15a4b0773adb8d5160a34d9a8e2d981394484.

* Not restoring spec/reactors/consensus/proposer-selection.md: already present under spec/consensus.
* Not restoring block_sync_img/bc-reactor.png: already present under docs/references/architecture/tendermint-core/img.
* Not restoring mempool/concurrency.md as it is a TODO list that needs to be documented.
* Not restoring spec/reactors/mempool/config.md as it looks like an excerpt from the reference manual. We keep the reference manual up-to-date instead.

Restoring

* spec/reactors/block_sync/img/bc-reactor-routines.png spec/reactors/consensus/consensus.md
* spec/reactors/consensus/consensus-reactor.md
* spec/reactors/mempool/functionality.md
* spec/reactors/block_sync/impl.md
* spec/reactors/mempool/messages.md
* spec/reactors/*/reactor.ms
* spec/reactors/readme.md (updated with legacy note)

---

These files are added (with a note that they are legacy documentation) as we did not find an updated version of them in our new folder structure. We should go through them, pick out the still relevant/useful information and decommission them for good.

### CODE REVIEW

**Comment by greg-szabo** [comment]
I identified bc-reactor.png as already present under architecture documentation and then I forgot to delete it from the commit. I'm open to suggestions to keep it or remove it.
Advantage of keeping it: it sits together with bc-reactor-routins.png which provides more context on what the file is about. When cleanup comes, it's easier to decide where to put both.
Disadvantage: duplicated file.

File: spec/reactors/readme.md
Comment (cason) [inline spec/reactors/readme.md:1]: Rename to `README.md`. Add a summary of the existing documentation in sub-dirs.

File: spec/reactors/block_sync/reactor.md [hunk 1]
```diff
@@ -0,0 +1,308 @@
+# Blockchain Reactor
```
Comment (cason) [inline spec/reactors/block_sync/reactor.md:1]:
```suggestion
# BlockSync Reactor
```

File: spec/reactors/block_sync/reactor.md
Comment (cason) [inline spec/reactors/block_sync/reactor.md:1]: I would rename it to `README.md`.

File: spec/reactors/block_sync/reactor.md [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+
+The Blockchain Reactor's high level responsibility is to enable peers who are
+far behind the current state of the consensus to quickly catch up by downloading
+many blocks in parallel, verifying their commits, and executing them against the
+ABCI application.
+
+Tendermint full nodes run the Blockchain Reactor as a service to provide blocks
```
Comment (cason) [inline spec/reactors/block_sync/reactor.md:8]:
```suggestion
CometBFT full nodes run the BlockSync Reactor as a service to provide blocks
```

File: spec/reactors/block_sync/reactor.md [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+to new nodes. New nodes run the Blockchain Reactor in "fast_sync" mode,
```
Comment (cason) [inline spec/reactors/block_sync/reactor.md:9]:
```suggestion
to new nodes. New nodes run the BlockSync Reactor in "block_sync" mode,
```

File: spec/reactors/block_sync/reactor.md [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+where they actively make requests for more blocks until they sync up.
+Once caught up, "fast_sync" mode is disabled and the node switches to
```
Comment (cason) [inline spec/reactors/block_sync/reactor.md:11]:
```suggestion
Once caught up, "block_sync" mode is disabled and the node switches to
```

File: spec/reactors/block_sync/reactor.md [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+using (and turns on) the Consensus Reactor.
+
+## Message Types
```
Comment (cason) [inline spec/reactors/block_sync/reactor.md:14]: We should replace by the protobuf version of them here.

File: spec/reactors/block_sync/reactor.md [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+  Height int64
+  PeerID p2p.ID
+}
+```
+
+### Receive routine of Blockchain Reactor
+
+It is executed upon message reception on the BlockchainChannel inside p2p receive routine. There is a separate p2p receive routine (and therefore receive routine of the Blockchain Reactor) executed for each peer. Note that try to send will not block (returns immediately) if outgoing buffer is full.
+
+```go
```
Comment (cason) [inline spec/reactors/block_sync/reactor.md:114]: Too much low level, I'd say.

File: spec/reactors/block_sync/impl.md
Comment (cason) [inline spec/reactors/block_sync/impl.md:1]: We only have one implementation right now. I would say to merge this document into the main one.

File: spec/reactors/evidence/reactor.md
Comment (cason) [inline spec/reactors/evidence/reactor.md:1]: Here we should link to the evidence production and validation specs.

File: spec/reactors/evidence/reactor.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# Evidence Reactor
+
+## Channels
+
+[#1503](https://github.com/tendermint/tendermint/issues/1503)
```
Comment (cason) [inline spec/reactors/evidence/reactor.md:5]: This is closed, we can remove.

File: spec/reactors/mempool/reactor.md
Comment (cason) [inline spec/reactors/mempool/reactor.md:1]: This is super outdated, we don't use batches any longer. We also have a `spec/mempool` directory, we more updated documentation.

File: spec/reactors/pex/pex.md
Comment (cason) [inline spec/reactors/pex/pex.md:1]: This should be merged into the content in `spec/p2p/implementation/pex.md`.

File: spec/reactors/mempool/messages.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+the p2p gossip network (via the reactor): `TxsMessage`
+
+```go
+// TxsMessage is a MempoolMessage containing a list of transactions.
+type TxsMessage struct {
+    Txs []types.Tx
+}
+```
+
+## RPC Messages
```
Comment (cason) [inline spec/reactors/mempool/messages.md:15]: This is more documentation than spec, I'd say.

We can check the related content there to merge.

File: spec/reactors/mempool/functionality.md
Comment (cason) [inline spec/reactors/mempool/functionality.md:1]: This is a good summary, but for `spec/mempool`.

File: spec/reactors/state_sync/reactor.md
Comment (cason) [inline spec/reactors/state_sync/reactor.md:1]: Also a neat summary, it should go somewhere else maybe.

File: spec/reactors/state_sync/reactor.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  network.
+
+* Discovering existing snapshots and fetching snapshot chunks for an empty local application
+  being bootstrapped.
+
+The state sync process for bootstrapping a new node is described in detail in the section linked
+above. While technically part of the reactor (see `statesync/syncer.go` and related components),
+this document will only cover the P2P reactor component.
+
+For details on the ABCI methods and data types, see the [ABCI documentation](../../abci/abci.md).
```
Comment (cason) [inline spec/reactors/state_sync/reactor.md:18]: Broken link.

File: spec/reactors/consensus/proposer-selection.md
Comment (cason) [inline spec/reactors/consensus/proposer-selection.md:1]: This file is identical to `spec/consensus/proposer-selection.md`, we should drop it.

File: spec/reactors/mempool/config.md
Comment (cason) [inline spec/reactors/mempool/config.md:1]: This is redundant with the documentation, also outdated.

File: spec/reactors/mempool/concurrency.md
Comment (cason) [inline spec/reactors/mempool/concurrency.md:1]: This is a bit TODO. Nice to have done, but not providing useful information with this content.

File: spec/reactors/consensus/consensus-reactor.md
Comment (cason) [inline spec/reactors/consensus/consensus-reactor.md:1]: This is probably mostly accurate. We can work a little on it and render it a relevant document.

File: spec/reactors/consensus/consensus.md
Comment (cason) [inline spec/reactors/consensus/consensus.md:1]: This file is confusing. We are handling reactor level messages in the consensus description. Not great, but something can be re-used here to produce a better documentation.

**Review by cason — COMMENTED** [review]
So, there are parts of this documentation that are relevant, no questions about it. We need to do a short review on naming (Tendermint -> Comet) and other reactors that have changed names (e.g., fast sync/blockchain -> block sync).

I would suggest, however, to call the base directory `spec/legacy` instead of `spec/reactors`. And make it very clear that this is old documentation on the `README.md` entry file of this documentation, and also in all main files (`README.md`) on each directory.

**Comment by cason** [comment]
> Not restoring spec/reactors/consensus/proposer-selection.md: already present under spec/consensus.

This file is still here.

**Comment by cason** [comment]
@greg-szabo , should I take this over?

**Comment by melekes** [comment]
> @greg-szabo , should I take this over?

please do @cason 🙏

**Comment by melekes** [comment]
Approving so we can merge this. We can address @cason comments https://github.com/cometbft/cometbft/pull/3494#pullrequestreview-2174276751 in a separate PR.

**Pull request merged**